### PR TITLE
fix(qsa): prevent out-of-bounds reads from padded extend plans

### PR DESCRIPTION
**Pull request by Wei-Nijuan**
## Motivation

The fixed-capacity QSA compressed-write plan pads unused entries with row 0
and a reserved slot-0 write. The write is inert, but the corresponding extend
read is not: `member_rows=0` is expanded to `[0, 1, 2, 3]` for a compression
ratio of 4.

An extend shorter than the compression ratio has no complete group, but its
shape-derived plan still contains one padded entry. A one-token extend
therefore reads rows 1 through 3 from a one-row `token_k` tensor. Depending on
the allocation layout, this can surface as an index error or a CUDA illegal
memory access.

This is a follow-up to the short-extend out-of-bounds report in #36497.

## Modifications

- Carry the write plan's `valid` mask through `QSAIndexerMetadata`.
- Before eager or fused extend compression, fold every padded entry's member
  locations onto its in-bounds dummy row.
- Preserve the fixed-capacity plan, reserved slot-0 write convention, and
  device-only/no-host-sync execution.
- Add regression coverage for 1-, 2-, and 3-token extends and for a real
  compression group followed by a padded entry.

## Accuracy Tests

- The new short-extend regression reproduces an `IndexError` on the unpatched
  source and passes for extend lengths 1, 2, and 3 after the fix.
- A four-token real group still produces the same FP32 average while the
  trailing padded entry writes only to reserved slot 0.
- `black`, `isort`, Python compilation, and `git diff --check` pass for all
  modified files.

Full registered GPU CI is left to the upstream workflow.

## Speed Tests and Profiling

Not separately profiled. The change preserves all plan shapes and adds no host
synchronization; it adds one elementwise selection over the fixed-capacity
extend compression plan before both eager and fused compression.

## Checklist

- [x] Format the modified code with the repository's Black and isort versions.
- [x] Add focused unit regression coverage.
- [x] No documentation update is needed for this correctness fix.
- [ ] Run the upstream registered GPU test suite.
- [x] Follow the existing SGLang code style.

---
### CI States

Latest PR Test (Base): :x: [Run #34829431806](https://github.com/sgl-project/sglang/actions/runs/34829431806)
Latest PR Test (Extra): :x: [Run #34829431456](https://github.com/sgl-project/sglang/actions/runs/34829431456)
Latest PR Test (AMD ROCm 10): :x: [Run #34829431756](https://github.com/sgl-project/sglang/actions/runs/34829431756)

### CODE REVIEW

**Comment by Leslie360** [comment]
## Independent SM80 (8×A800) verification

Independent verification of this PR on a single node of 8×A800 80GB (SM80), CUDA 13.0, torch 2.11.0+cu130, FA2 2.8.3.post1, TP8, running sglang from source (qwen4 branch, base commit `56bba54151b6922e18f4150b2c0583042df65adb`).

### Patch application
- `git apply --check` on a clean baseline: **OK** (forward, reverse, and whitespace all clean; no manual backport needed). 4 files, +105/−13.

### Serve + short-extend stress (base vs patched, identical loads)
Serving args: `--tp 8 --context-length 40000 --chunked-prefill-size 2048 --mem-fraction-static 0.80`. Both `/health` return 200.

| Scenario | Load | base (unpatched) | patched |
|---|---|---|---|
| Sequential short extend (1–6 token prompts, i.e. extend < compress_ratio=4, padded) | 150 req | ok=150, fail=0 | ok=150, fail=0 |
| Concurrent short extend (1–6 token prompts) | 400 req, conc 32 | ok=400, fail=0 (10.0s) | ok=400, fail=0 (9.0s) |
| Chunk-boundary (prompts ~7K tokens whose trailing chunk is 1–4 tokens) | 24 req | ok=24, fail=0 | ok=24, fail=0 |
| 8K long-context (patched) | 11,138 prompt tokens | — | OK in 2.3s, needle correctly retrieved |

Observation: on this SM80/TP8 platform the padded-extend access is a *read* OOB that typically stays within the packed tensor allocation; the garbage lands in masked positions, so the unpatched build did not show a user-visible crash in any stress scenario. The fix still closes a genuine latent OOB-read hazard (directly exercised by the PR's new unit tests) and is behaviorally neutral.

### Regression: GSM8K-50 (temperature 0), per-question comparison
- base: 47/50 (0.9400); patched: 48/50 (0.9600)
- Per-question prediction agreement: **48/50 identical**. The 2 differing questions (idx 12, idx 40) have different full outputs in both builds and are within expected MoE fp-non-determinism at temp=0 (idx 12: both wrong with different wrong answers; idx 40: base wrong, patched correct — patched not worse in either case). **No regression.**

### Unit tests (patched)
- New tests `test_qsa_short_extend_padding_uses_in_bounds_dummy_reads[1/2/3]` and `test_qsa_extend_padding_does_not_change_real_group_compression`: **PASSED**.
- Full `test/registered/kernels/test_qsa.py`: **72 passed, 0 failed** (existing QSA kernel tests unaffected).

### Verdict
Supports merge. Patch applies cleanly to the current head, serves stably on SM80 TP8 across short/concurrent/chunk-boundary and 8K long-context, unit tests pass, and GSM8K-50 shows no output regression versus the unpatched baseline. On our platform we could not reproduce a pre-fix crash (the OOB read is masked in practice), but the fix correctly and safely closes the underlying OOB-read defect.

**Comment by chenweigao** [comment]
# additional SM120 evidence for QSA padded-extend OOB

Suggested destination: https://github.com/sgl-project/sglang/pull/37786
Related model-support discussion: https://github.com/sgl-project/sglang/pull/36497
Status: local draft, not posted.

We independently reproduced the padded compressed-write-plan out-of-bounds read described here, on RTX PRO 5000 (SM120). This adds a Compute Sanitizer reproducer for the fused CUDA kernel.

## Environment

- GPU used by the isolated reproducer: one RTX PRO 5000, 72 GB, SM120.
- Serving setup where we investigated: four GPUs, TP2 DP2, RadixArk/Qwen3.8-Flash-Next-NVFP4, NEXTN steps=3/draft tokens=4, PLE offload, chunked prefill 4096, CUDA Graph enabled for decode.
- Cookbook image: lmsysorg/sglang:dev-qwen38-next; tested amd64 image digest sha256:9d278f74612e692c90a7ebbb6eb820eed95e2ffd321e62c75fd94a6141ceef74.
- Source revision recorded in image: 4ccff141dbe992794f9da6c3aa23535b4f72000d on qwen4-main-squashed.
- torch 2.13 / CUDA 13.0, FlashInfer 0.6.17.
- Serving also has a local scheduler change alternating separate prefill and decode batches. The isolated kernel reproducer below does not load the scheduler or model and is independent of that change.

## Reproducer (no model weights required)

Save as repro.py in the affected image and run:

```bash
PYTORCH_NO_CUDA_MEMORY_CACHING=1 compute-sanitizer --tool memcheck --error-exitcode 9 python repro.py
```

```python
"""Reproduce the QSA write-plan padding group on a short cached extension."""
import torch
from sglang.kernels.ops.attention.qsa_indexer import qsa_index_k_compress_store

device = 'cuda'
keys = torch.ones((1, 128), dtype=torch.bfloat16, device=device)
group_locs = torch.arange(4, dtype=torch.int32, device=device).reshape(1, 4)
positions = torch.zeros((1, 3), dtype=torch.int64, device=device)
cache = torch.cat((torch.ones(8, 16), torch.zeros(8, 16)), dim=1).to(device)
axis = torch.zeros(16, dtype=torch.int32, device=device)
weight = torch.zeros(128, dtype=torch.bfloat16, device=device)
write_locs = torch.zeros(1, dtype=torch.int32, device=device)
output = torch.zeros((1, 128), dtype=torch.bfloat16, device=device)
qsa_index_k_compress_store(keys, group_locs, positions, cache, axis,
                          weight, write_locs, output, 4, 32, 1e-6, True)
torch.cuda.synchronize()
print('kernel completed')
```

This directly constructs the padded entry produced by a one-token extend: compress_ratio=4, member start zero, write_locs=[0]. No complete group should read source rows 1..3.

## Observed evidence

Compute Sanitizer reported 36 errors on the original kernel, beginning with an invalid 8-byte global read in qsa_index_k_compress_kernel<bf16,128,true,true>. The first address was just beyond the one-row, 256-byte key allocation.

A local alternative fix skips entries whose write_locs[group]==0 after PDLWaitPrimary and before reading group members. Under Compute Sanitizer, this fix passed three short-extend regressions (1, 2, 3 tokens) and nine valid-group comparisons against the eager reference: 12 passed, ERROR SUMMARY: 0 errors. The numerical fixture needed updating for the current _rope_from_matrix helper and index_state_dtype field; real write targets excluded reserved slot zero.

This local kernel-only fix is NOT a validation of the full patch in this PR; your validity-mask approach covers both eager and fused consumers. Our local fix covers the fused consumer we use. We are sharing the independently observed failure and hardware evidence rather than proposing a duplicate PR.

## Serving observations and limitations

Before the local fix, three serving crashes surfaced in QSA get_prefill_mqa_inputs at sequence_lengths.tolist(), with CUDA illegal memory access. The precise original failing requests were not captured. Targeted 7040-token cached prefixes plus 1..4 new tokens returned HTTP 200 even on the original kernel, so we do not claim a deterministic API-level crash reproducer or that this defect explains every serving crash.

After deploying the local kernel fix and removing temporary synchronization diagnostics, six concurrent arithmetic checks, a real OMP plain SSE request, a real OMP tool-call SSE request, and a vision request all passed (9/9). Long-running stability remains an observation, not proof of complete incident attribution.

**Comment by ykamakazi** [comment]
ROCm data point for this bug (deterministic reproducer + before/after on 1,500 requests) posted on #38346: https://github.com/sgl-project/sglang/pull/38346#issuecomment-5621691108 — same conclusion applies to this PR.
